src/awkward/_v2/operations/describe/ak_type.py

```diff
@@ -2,147 +2,154 @@
-#             for x in array.shape[-1:0:-1]:
-#                 out = ak._v2.types.RegularType(out, x)
-#             return ak._v2.types.ArrayType(out, array.shape[0])
-
-#     elif isinstance(array, ak.layout.ArrayBuilder):
-#         return array.type(ak._v2._util.typestrs(None))
-
-#     elif isinstance(array, (ak._v2.contents.Content, ak.partition.PartitionedArray)):   # NO PARTITIONED ARRAY
-#         return array.type(ak._v2._util.typestrs(None))
-
-#     else:
-#         raise TypeError(
-#             "unrecognized array type: {0}".format(repr(array))
-#
-#         )
-
-
-# type.dtype2primitive = {
-#     np.bool_: "bool",
-#     np.int8: "int8",
-#     np.int16: "int16",
-#     np.int32: "int32",
-#     np.int64: "int64",
-#     np.uint8: "uint8",
-#     np.uint16: "uint16",
-#     np.uint32: "uint32",
-#     np.uint64: "uint64",
-#     np.float32: "float32",
-#     np.float64: "float64",
-#     np.complex64: "complex64",
-#     np.complex128: "complex128",
-#     np.datetime64: "datetime64",
-#     np.timedelta64: "timedelta64",
-# }
+    """
+    The high-level type of an `array` (many types supported, including all
+    Awkward Arrays and Records) as #ak.types.Type objects.
+
+    The high-level type ignores #layout differences like
+    #ak.layout.ListArray64 versus #ak.layout.ListOffsetArray64, but
+    not differences like "regular-sized lists" (i.e.
+    #ak.layout.RegularArray) versus "variable-sized lists" (i.e.
+    #ak.layout.ListArray64 and similar).
+
+    Types are rendered as [Datashape](https://datashape.readthedocs.io/)
+    strings, which makes the same distinctions.
+
+    For example,
+
+        ak.Array([[{"x": 1.1, "y": [1]}, {"x": 2.2, "y": [2, 2]}],
+                  [],
+                  [{"x": 3.3, "y": [3, 3, 3]}]])
+
+    has type
+
+        3 * var * {"x": float64, "y": var * int64}
+
+    but
+
+        ak.Array(np.arange(2*3*5).reshape(2, 3, 5))
+
+    has type
+
+        2 * 3 * 5 * int64
+
+    Some cases, like heterogeneous data, require [extensions beyond the
+    Datashape specification](https://github.com/blaze/datashape/issues/237).
+    For example,
+
+        ak.Array([1, "two", [3, 3, 3]])
+
+    has type
+
+        3 * union[int64, string, var * int64]
+
+    but "union" is not a Datashape type-constructor. (Its syntax is
+    similar to existing type-constructors, so it's a plausible addition
+    to the language.)
+    """
+    if array is None:
+        return ak._v2.types.UnknownType()
+
+    elif isinstance(
+        array,
+        (
+            np.int8,
+            np.int16,
+            np.int32,
+            np.int64,
+            np.uint8,
+            np.uint16,
+            np.uint32,
+            np.uint64,
+            np.float32,
+            np.float64,
+            np.complex64,
+            np.complex128,
+            np.datetime64,
+            np.timedelta64,
+        ),
+    ):
+        return ak._v2.types.PrimitiveType(type.dtype2primitive[array.dtype.type])
```

Comment:
```suggestion
        return ak._v2.types.PrimitiveType(ak._v2.types.numpytype._dtype_to_primitive_dict[array.dtype])
```

Both this and the above change have to go together. I should have made it one edit.